src/Pool.sol

```diff
@@ -378,49 +369,44 @@ abstract contract Pool {
     }
 
     /// @notice Updates all the sender's parameters.
-    /// Tops up and withdraws unsent funds from the balance of the sender.
+    /// Transfers funds to or from the sender to fulfill the update of the balance.
     /// @param senderId The sender id
     /// @param lastUpdate The timestamp of the last update of the sender.
     /// If this is the first update of the sender, pass zero.
     /// @param lastBalance The balance after the last update of the sender.
     /// If this is the first update of the sender, pass zero.
     /// @param currReceivers The list of receivers set in the last update of the sender.
     /// If this is the first update of the sender, pass an empty array.
-    /// @param topUpAmt The topped up amount.
-    /// @param withdrawAmt The amount to be withdrawn, must not be higher than available funds.
-    /// Can be `WITHDRAW_ALL` to withdraw everything.
+    /// @param balanceDelta The sender balance change to be applied.
     /// @param newReceivers The new list of the sender's receivers.
     /// Must be sorted by the receivers' addresses, deduplicated and without 0 amtPerSecs.
     /// @return newBalance The new sender balance.
     /// Pass it as `lastBalance` when updating the user for the next time.
-    /// @return withdrawn The actually withdrawn amount.
-    /// Equal to `withdrawAmt` unless `WITHDRAW_ALL` has been used.
+    /// @return realBalanceDelta The actually applied balance change.
     function _updateSender(
         SenderId memory senderId,
         uint64 lastUpdate,
         uint128 lastBalance,
         Receiver[] calldata currReceivers,
-        uint128 topUpAmt,
-        uint128 withdrawAmt,
+        int128 balanceDelta,
         Receiver[] calldata newReceivers
-    ) internal returns (uint128 newBalance, uint128 withdrawn) {
+    ) internal returns (uint128 newBalance, int128 realBalanceDelta) {
         _assertSenderState(senderId, lastUpdate, lastBalance, currReceivers);
         uint128 newAmtPerSec = _assertReceiversValid(newReceivers);
         uint128 currAmtPerSec = _totalAmtPerSec(currReceivers);
         uint64 currEndTime = _sendingEndTime(lastUpdate, lastBalance, currAmtPerSec);
-        (newBalance, withdrawn) = _updateSenderBalance(
+        (newBalance, realBalanceDelta) = _updateSenderBalance(
             lastUpdate,
             lastBalance,
             currEndTime,
-            topUpAmt,
-            withdrawAmt,
-            currAmtPerSec
+            currAmtPerSec,
+            balanceDelta
         );
         uint64 newEndTime = _sendingEndTime(_currTimestamp(), newBalance, newAmtPerSec);
         _updateStreams(senderId, currReceivers, currEndTime, newReceivers, newEndTime);
         _storeSenderState(senderId, newBalance, newReceivers);
         _emitSenderUpdated(senderId, newBalance, newReceivers);
-        _transfer(senderId.senderAddr, int128(withdrawn) - int128(topUpAmt));
+        _transfer(senderId.senderAddr, -realBalanceDelta);
```

Comment:
`int128(currBalance-balance)` instead of `int128(balance - currBalance)`

Reply:
balanceDelta => positive => topUp
transferDelta => positive => withdraw

I think we maybe should just change the `tranfer` delta to positive is topUp

Reply:
We still need to return `realBalanceDelta` as it is, following the semantics of `balanceDelta`. If we make `realBalanceDelta = -realBalanceDelta`, we'll need to negate it once more before returning.

Let's not introduce more variable semantics, introducing a `transferDelta` with a new meaning is not worth removing the minus sign from `_transfer`. I wouldn't touch `_transfer` either, it's used in other places like `collect`, where it accepts a positive value.

Reply:
> Let's not introduce more variable semantics, introducing a transferDelta with a new meaning
I think it would simplify the codebase if we do it the same way in `_transfer`.  

I agree with the `realBalanceDelta` and `balanceDelta` that they should have the same behaviour.

For me the starting point is the EthPool here  msg.value is positive indicate a `topUp`. This we can't change.
I think `_transfer` should act the same. Obviously it means we would have to flip the signs at the usage.

Reply:
Remember that `_transfer` is used not only in `updateSender`, but also in `collect`. If we change `_transfer` to `_transferFromContract`, it would move the dreaded minus sign to `collect`. It'd also add minus signs to implementations of `_transfer`, because ERC20's and Etehreum's `transfer` don't work like that.

Reply:
I can see your reasoning, it is not a merge blocker for me.